File: backend/src/controllers/community.controller.ts
import type { Request, Response } from "express";
import { Types } from "mongoose";
import { communityService } from "../services/community.service.js";
import { ApiResponse } from "../utils/api-response.js";
import { ApiError } from "../utils/api-error.js";
import { Community, type IChannel } from "../models/community.model.js";
import { AdminLog } from "../models/admin-log.model.js";
import { getSocketServer } from "../sockets/index.js";
import type {
  CreateCommunityInput,
  ListCommunitiesQuery,
  UpdateCommunityInput
} from "../validators/community.validator.js";

export class CommunityController {
  async create(request: Request, response: Response) {
    const community = await communityService.create(
      request.body as CreateCommunityInput,
      request.user!.id,
      request.file
    );
    response.status(201).json(new ApiResponse(201, community, "Community created"));
  }

  async list(request: Request, response: Response) {
    const communities = await communityService.list(
      request.validated?.query as ListCommunitiesQuery,
      request.user!.id
    );
    response.json(new ApiResponse(200, communities, "Communities retrieved"));
  }

  async details(request: Request, response: Response) {
    const community = await communityService.details(param(request, "id"), request.user!.id);
    response.json(new ApiResponse(200, community, "Community retrieved"));
  }

  async update(request: Request, response: Response) {
    const community = await communityService.update(
      param(request, "id"),
      request.body as UpdateCommunityInput,
      request.user!.id,
      request.file
    );
    response.json(new ApiResponse(200, community, "Community updated"));
  }

  async delete(request: Request, response: Response) {
    await communityService.delete(param(request, "id"), request.user!.id);
    response.json(new ApiResponse(200, null, "Community deleted"));
  }

  async join(request: Request, response: Response) {
    const community = await communityService.join(param(request, "id"), request.user!.id);
    response.json(new ApiResponse(200, community, "Joined community"));
  }

  async leave(request: Request, response: Response) {
    await communityService.leave(param(request, "id"), request.user!.id);
    response.json(new ApiResponse(200, null, "Left community"));
  }

  async members(request: Request, response: Response) {
    const members = await communityService.membersList(param(request, "id"), request.user!.id);
    response.json(new ApiResponse(200, members, "Community members retrieved"));
  }

  async addModerator(request: Request, response: Response) {
    const members = await communityService.addModerator(
      param(request, "id"),
      request.body.userId,
      request.user!.id
    );
    response.json(new ApiResponse(200, members, "Moderator added"));
  }

  async removeModerator(request: Request, response: Response) {
    const members = await communityService.removeModerator(
      param(request, "id"),
      param(request, "userId"),
      request.user!.id
    );
    response.json(new ApiResponse(200, members, "Moderator removed"));
  }

  async removeMember(request: Request, response: Response) {
    const members = await communityService.removeMember(
      param(request, "id"),
      param(request, "userId"),
      request.user!.id
    );
    response.json(new ApiResponse(200, members, "Member removed"));
  }

  async updateChannelStudyMode(request: Request, response: Response) {
    const communityId = param(request, "id");
    const channelId = param(request, "channelId");
    const userId = request.user!.id;
    const userRole = request.user!.role;

    const community = await Community.findById(communityId);
    if (!community) {
      throw new ApiError(404, "Community not found", [], "COMMUNITY_NOT_FOUND");
    }

    const isOwner = community.owner.toString() === userId;
    const isModerator = community.moderators.some((m) => m.toString() === userId);
    const isAdmin = userRole === "ADMIN";

    if (!isOwner && !isModerator && !isAdmin) {
      throw new ApiError(
        403,
        "Only verified room moderators or administrators can configure study mode",
        [],
        "FORBIDDEN_STUDY_MODE"
      );
    }

    const {
      isStrictStudyMode,
      academicContextTags,
      strictnessThreshold,
      allowCodeSnippetsOnly,
      strikeLimitBeforeTimeout,
      timeoutDurationMinutes
    } = request.body;

    let targetChannel = community.channels.find(
      (c) => c._id?.toString() === channelId || c.name.toLowerCase() === channelId.toLowerCase()
    );

    if (!targetChannel) {
      const newChan: IChannel = {
        name: channelId,
        type: "text",
        isStrictStudyMode: isStrictStudyMode ?? true,
        academicContextTags: Array.isArray(academicContextTags)
          ? academicContextTags
          : ["algorithms", "code", "homework", "exam"],
        strictnessThreshold:
          typeof strictnessThreshold === "number" ? strictnessThreshold : 0.4,
        allowCodeSnippetsOnly: Boolean(allowCodeSnippetsOnly),
        strikeLimitBeforeTimeout: strikeLimitBeforeTimeout ?? 3,
        timeoutDurationMinutes: timeoutDurationMinutes ?? 5
      };
      community.channels.push(newChan);
      targetChannel = newChan;
    } else {
      if (typeof isStrictStudyMode === "boolean") targetChannel.isStrictStudyMode = isStrictStudyMode;
      if (Array.isArray(academicContextTags)) targetChannel.academicContextTags = academicContextTags;
      if (typeof strictnessThreshold === "number") targetChannel.strictnessThreshold = strictnessThreshold;
      if (typeof allowCodeSnippetsOnly === "boolean") targetChannel.allowCodeSnippetsOnly = allowCodeSnippetsOnly;
      if (typeof strikeLimitBeforeTimeout === "number") targetChannel.strikeLimitBeforeTimeout = strikeLimitBeforeTimeout;
      if (typeof timeoutDurationMinutes === "number") targetChannel.timeoutDurationMinutes = timeoutDurationMinutes;
    }

    await community.save();

    // Log admin / mod audit trail
    try {
      await AdminLog.create({
        adminId: new Types.ObjectId(userId),
        adminName: request.user?.fullName || "Room Moderator",
        action: "UPDATE_CHANNEL_STUDY_MODE",
        targetType: "Channel",
        targetId: channelId,
        details: {
          communityId,
          channelId,
          isStrictStudyMode: targetChannel?.isStrictStudyMode,
          academicContextTags: targetChannel?.academicContextTags,
          strictnessThreshold: targetChannel?.strictnessThreshold,
          allowCodeSnippetsOnly: targetChannel?.allowCodeSnippetsOnly
        }
      });
    } catch {}

    // Emit live socket event to all active room subscribers
    try {
      const io = getSocketServer();
      if (io) {
        const payload = {
          communityId,
          channelId,
          channel: targetChannel
        };
        io.to(`room:${communityId}`).emit("channel:studyModeUpdated", payload);
        io.to(`room:${communityId}:${channelId}`).emit("channel:studyModeUpdated", payload);
      }
    } catch {}

    response.json(new ApiResponse(200, targetChannel, "Channel study mode configuration updated"));
  }
}

export const communityController = new CommunityController();

const param = (request: Request, key: string): string => request.params[key] as string;

File: backend-java/src/main/java/com/studyconnect/backend/controller/CommunityController.java
package com.studyconnect.backend.controller;

import com.studyconnect.backend.dto.ApiResponse;
import com.studyconnect.backend.dto.community.CommunityCreateRequest;
import com.studyconnect.backend.dto.community.CommunityDto;
import com.studyconnect.backend.dto.community.CommunityMemberDto;
import com.studyconnect.backend.dto.community.CommunityPageDto;
import com.studyconnect.backend.dto.community.CommunityUpdateRequest;
import com.studyconnect.backend.dto.community.ModeratorRequest;
import com.studyconnect.backend.service.CommunityService;
import com.studyconnect.backend.util.SecurityUtil;
import jakarta.validation.Valid;
import jakarta.validation.constraints.Max;
import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.Size;
import java.util.List;
import org.springframework.security.core.Authentication;
import org.springframework.validation.annotation.Validated;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.ModelAttribute;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

@Validated
@RestController
@RequestMapping("/api/communities")
public class CommunityController {

    private final CommunityService communityService;

    public CommunityController(CommunityService communityService) {
        this.communityService = communityService;
    }

    @PostMapping(consumes = "multipart/form-data")
    public ResponseEntity<ApiResponse<CommunityDto>> create(@Valid @ModelAttribute CommunityCreateRequest request, Authentication authentication) {
        return ResponseEntity.status(201).body(ApiResponse.success(communityService.create(request, SecurityUtil.currentUserId(authentication)), "Community created"));
    }

    @GetMapping
    public ApiResponse<CommunityPageDto> list(
            @RequestParam(name = "search", required = false) @Size(max = 100) String search,
            @RequestParam(name = "category", required = false) String category,
            @RequestParam(name = "page", defaultValue = "1") @Min(1) int page,
            @RequestParam(name = "limit", defaultValue = "12") @Min(1) @Max(50) int limit,
            Authentication authentication) {
        return ApiResponse.success(communityService.list(search, category, page, limit, SecurityUtil.currentUserId(authentication)), "Communities retrieved");
    }

    @GetMapping("/{id}")
    public ApiResponse<CommunityDto> details(@PathVariable String id, Authentication authentication) {
        return ApiResponse.success(communityService.details(id, SecurityUtil.currentUserId(authentication)), "Community retrieved");
    }

    @PutMapping(value = "/{id}", consumes = "multipart/form-data")
    public ApiResponse<CommunityDto> update(
            @PathVariable String id,
            @Valid @ModelAttribute CommunityUpdateRequest request,
            Authentication authentication) {
        return ApiResponse.success(communityService.update(id, request, SecurityUtil.currentUserId(authentication)), "Community updated");
    }

    @DeleteMapping("/{id}")
    public ApiResponse<Void> delete(@PathVariable String id, Authentication authentication) {
        communityService.delete(id, SecurityUtil.currentUserId(authentication));
        return ApiResponse.success(null, "Community deleted");
    }

    @PostMapping("/{id}/join")
    public ApiResponse<CommunityDto> join(@PathVariable String id, Authentication authentication) {
        return ApiResponse.success(communityService.join(id, SecurityUtil.currentUserId(authentication)), "Joined community");
    }

    @PostMapping("/{id}/leave")
    public ApiResponse<Void> leave(@PathVariable String id, Authentication authentication) {
        communityService.leave(id, SecurityUtil.currentUserId(authentication));
        return ApiResponse.success(null, "Left community");
    }

    @GetMapping("/{id}/members")
    public ApiResponse<List<CommunityMemberDto>> members(@PathVariable String id, Authentication authentication) {
        return ApiResponse.success(communityService.membersList(id, SecurityUtil.currentUserId(authentication)), "Community members retrieved");
    }

    @PostMapping("/{id}/moderators")
    public ApiResponse<List<CommunityMemberDto>> addModerator(
            @PathVariable String id,
            @Valid @RequestBody ModeratorRequest request,
            Authentication authentication) {
        return ApiResponse.success(communityService.addModerator(id, request.userId(), SecurityUtil.currentUserId(authentication)), "Moderator added");
    }

    @DeleteMapping("/{id}/moderators/{userId}")
    public ApiResponse<List<CommunityMemberDto>> removeModerator(
            @PathVariable String id,
            @PathVariable String userId,
            Authentication authentication) {
        return ApiResponse.success(communityService.removeModerator(id, userId, SecurityUtil.currentUserId(authentication)), "Moderator removed");
    }

    @DeleteMapping("/{id}/members/{userId}")
    public ApiResponse<List<CommunityMemberDto>> removeMember(
            @PathVariable String id,
            @PathVariable String userId,
            Authentication authentication) {
        return ApiResponse.success(communityService.removeMember(id, userId, SecurityUtil.currentUserId(authentication)), "Member removed");
    }
}
